userspace/libsinsp/threadinfo.cpp

```diff
@@ -149,9 +149,9 @@ void sinsp_threadinfo::init()
 	m_exe_upper_layer = false;
 	m_exe_from_memfd = false;
 
-	memset(&m_user, 0, sizeof(scap_userinfo));
-	memset(&m_group, 0, sizeof(scap_groupinfo));
-	memset(&m_loginuser, 0, sizeof(scap_userinfo));
+	m_user = std::make_shared<sinsp_userinfo>();
```

Comment:
What do you mean? If `m_user` and friends are changed to be shared pointers, then surely this change is needed.

Reply:
Oh i meant to make them shared pointers, sorry!

Reply:
I see, thanks. I think making them a some sort of pointer is important to achiever minimal memory footprint in case if user_details is disabled, and having said that it's only reasonable to make them also a smart pointer. Unless there are any strong objection (?), I would keep it that way.

Reply:
Oh you are right, makes sense thank you!
Also, leveraging your idea of letting them become a class and attaching methods (https://github.com/falcosecurity/libs/pull/1765#discussion_r1542600598), we could make all methods return `<NA>` also when `this == nullptr`, ie: when the method is called on a null class.
This way, we could still easily call `tinfo->m_user->name()` even if `m_user` is nullptr, without further checks.
Wdyt? I think this might be indeed a very good and clean solution, if you agree!

Reply:
> when this == nullptr, ie: when the method is called on a null class.

From what I understand, if `this == nullptr` and we try to call any method on it, the specification says it's an undefined behaviour. Which means it doesn't make sense to defend against this, and any examples would be a compilation warning I guess.

> if (this == nullptr) { return -1; }

I was in fact pondering about this, but from the different perspective. Currently `uid` is `unit32_t`, so an unsigned value gives no chances to use `-1` as an "invalid value". This in turn means we can't really tell apart a non-initialized value from a real one.

Maybe it makes sense to convert this field into a signed one, reserving `-1` as an invalid value. But if you don't mind, I would let it outside of this PR and pursue this in a follow-up.

Reply:
> the specification says it's an undefined behaviour

Oh, i was pretty sure it was allowed; sorry i mixed up go and c++ :) 

> Maybe it makes sense to convert this field into a signed one, reserving -1 as an invalid value. But if you don't mind, I would let it outside of this PR and pursue this in a follow-up.

Agree!